### PR TITLE
Added horizontal pod autoscaler example.

### PR DESCRIPTION
**Pull request by jszczepkowski**
Added horizontal pod autoscaler example.

### CODE REVIEW

**Comment by jszczepkowski** [comment]
CC @derekwaynecarr @fgrzadkowski

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Horizontal Pod Autoscaler
+
+This example presents horizontal autoscaling of pods: an experimental feature delivered in version 1.1 of Kubernetes
```
Comment (derekwaynecarr) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: I would rephrase this.

Horizontal pod autoscaling is an experimental feature in Kubernetes 1.1.  It allows the number of pods in a replication controller or deployment to scale automatically based on observed CPU or memory usage.

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+which automatically sets the number of pods in a replication controller (or deployment), base on CPU / memory usage.
+
+
+## Prerequisites
+
+This example requires a running Kubernetes cluster in the version 1.1 with experimental API enabled on API server (```--runtime-config=experimental/v1alpha1=true```),
```
Comment (derekwaynecarr) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: This is no longer accurate.  See the changes here: https://github.com/kubernetes/kubernetes/pull/15024

There is now just a single flag to enable the feature.

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    name: php-apache
+    namespace: default
+  target:
+    quantity: 100m
+    resource: cpu
+```
+
+Please note that it sets the target CPU consumption to 100 mili-cores.
+
+We will create the autoscaller by executing the following command:
```
Comment (derekwaynecarr) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: autoscaler

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+## Step Three: Increase load
+
+Now, we will see how the autoscaler reacts on the increased load of the server.
+We will start an infinite loop of queries to our server (please run it in a different terminal):
+
+```console
+$ while true; do curl http://146.148.6.244; done
+```
+
+We may examine, how CPU load was increased by executing:
```
Comment (derekwaynecarr) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Comment on the latency of the measurement?  Maybe just say do a ```watch kubectl get hpa``` and you will see the observed load increase?

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+```console
+$ kubectl get hpa
+NAME         REFERENCE                                   TARGET     CURRENT    MINPODS   MAXPODS   AGE
+php-apache   ReplicationController/default/php-apache/   100m cpu   471m cpu   1         10        10m
+```
+
+In the case presented here, it bumped CPU consumption to 471 mili-cores.
```
Comment (derekwaynecarr) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: milli-cores (typo is in multiple places)

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+```console
+$ kubectl create -f ./examples/experimental/horizontal-pod-autoscaling/hpa-php-apache.yaml
+horizontalpodautoscaler "php-apache" created
+```
+
+We may check the current status of autoscaler by running:
+
+```console
+$ kubectl get hpa
```
Comment (derekwaynecarr) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: I think this will be changing soon:

Be aware depending on merge order:
https://github.com/kubernetes/kubernetes/pull/15206

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+The image can be found [here](image/).
+It defines [index.php](image/index.php) page which performs some CPU intensive computations.
+
+First, we will start a replication controller running the image and expose it as an external service:
+
+```console
+$ kubectl run php-apache --image=gcr.io/dev-jsz/php-apache --replicas=1
+replicationcontroller "php-apache" created
+
+$ kubectl expose rc php-apache --port=80 --type=LoadBalancer
```
Comment (derekwaynecarr) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Not all clusters support LoadBalancer out of the box.  A number of folks use the Vagrant cluster to kick the tires on examples.  Maybe change to say if your cluster supports load balancing, do what you present.  

If your cluster does not support an external load balancer, create a ClusterIP service and maybe use the ```/api/v1/proxy/namespaces/default/services/php-apache``` endpoint instead to drive traffic into the cluster?

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 1 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 7 above)
Comment (fgrzadkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: I agree with @derekwaynecarr. AFAIU we don't have to expose service as an external load balancer.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 2 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 3 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 5 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+We may examine, how CPU load was increased by executing:
+
+```console
+$ kubectl get hpa
+NAME         REFERENCE                                   TARGET     CURRENT    MINPODS   MAXPODS   AGE
+php-apache   ReplicationController/default/php-apache/   100m cpu   471m cpu   1         10        10m
+```
+
+In the case presented here, it bumped CPU consumption to 471 mili-cores.
+In affect, the replication controller was resized to 5 replicas:
```
Comment (fgrzadkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: s/In affect/As a result

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+$ kubectl get rc
+CONTROLLER   CONTAINER(S)   IMAGE(S)                    SELECTOR         REPLICAS   AGE
+php-apache   php-apache     gcr.io/dev-jsz/php-apache   run=php-apache   10         31m
+```
+
+## Step Four: Stop load
+
+We will finish our example by stopping the user load.
+We will terminate both infinite while loops sending requests to the server and verify the result state:
```
Comment (fgrzadkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Please surround "while" with ``` to for better readability.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 4 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

**Comment by fgrzadkowski** [comment]
As discussed offline I think that this example missing the following things:
* We need to set requests for RC to make sure that replicas will be able to use more than 100m cores
* We should mentioned that we plan to change the API (e.g. custom metrics, target relative to target)

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 6 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Yea, it is not merged yet, but I think it will be a part of 1.1. Done.

**Comment by davidopp** [comment]
There are some minor spelling/grammar things but it looks good overall. I agree you should not use external LoadBalancer.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 7 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

**Comment by jszczepkowski** [comment]
@derekwaynecarr @fgrzadkowski @davidopp 
All comments are applied. PTAL

**Comment by jszczepkowski** [comment]
> We should mentioned that we plan to change the API (e.g. custom metrics, target relative to target)

I added a general comment that the API is subject to change. I think that the detailed description of the changes is too low level for this doc and I'm going to put it the user-guide doc for horizontal pod autoscaler.

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Horizontal Pod Autoscaler
+
+Horizontal pod autoscaling is an experimental feature in Kubernetes 1.1.
+It allows the number of pods in a replication controller or deployment to scale automatically based on observed CPU or memory usage.
+<b>Please note that the current API is tentative and will be subject to change before a stable version is released.</b>
+
```
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: I would add a new paragraph at the end here that says "In this document we explain how this feature works by walking you through an example of enabling horizontal pod autoscaling with the php-apache server."

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+To demonstrate horizontal pod autoscaler we will use a custom docker image based on php-apache server.
+The image can be found [here](image/).
+It defines [index.php](image/index.php) page which performs some CPU intensive computations.
+
+First, we will start a replication controller running the image and expose it as an external service:
+```console
+$ kubectl create -f ./examples/experimental/horizontal-pod-autoscaling/rc-php-apache.yaml
+replicationcontrollers/php-apache
+
+$ kubectl expose rc php-apache --port=80 --type=LoadBalancer
```
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: I thought you were going to use ClusterIP instead of LoadBalancer?

Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Oh  I see, you added a comment later. Never mind.

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+
+$ curl -k -u admin:password https://146.148.6.215/api/v1/proxy/namespaces/default/services/php-apache/
+OK!
+```
+
+
+## Step Two: Create horizontal pod autoscaler
+
+Now that the server is running, we will create a horizontal pod autoscaler for it.
+To create it, we will use [hpa-php-apache.yaml](hpa-php-apache.yaml) file, which look like this:
```
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: s/we will use/we will use the/
s/which look like this/which looks like this/

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+  minReplicas: 1
+  scaleRef:
+    kind: ReplicationController
+    name: php-apache
+    namespace: default
+  target:
+    quantity: 100m
+    resource: cpu
+```
+
```
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: I would add a sentence here explaining what this does, something like "This will create a horizontal pod autoscaler that maintains between 1 and 10 replicas of the Pods controlled by the `php-apache` replication controller you created in the first step of these instructions. Roughly speaking, the horizontal autoscaler will increase and decrease the number of replicas (via the replication controller) so as to maintain an average CPU utilization across all Pods of 100 millicores. See \<reference to proposals/ doc here\> for more details on the algorithm."

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+Please be aware that it sets the target CPU consumption to 100 milli-cores, while in [rc-php-apache.yaml](rc-php-apache.yaml) each pod requests 200 milli-cores.
```
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: s/it/this configuration/
(the "it" will be ambiguous once you add the sentence I suggested in the previous comment)

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+As a general rule, the autoscaler's target should be lower than the request.
+Otherwise, overloaded pods may not be able to consume more resources than requested what will prevent scaling-up.
```
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: I would rewrite this sentence to be a little clearer: "Otherwise, overloaded pods may not be able to consume more than the autoscaler's target utilization, thereby preventing the autoscaler from seeing high enough utilization to trigger it to scale up."

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 13 above)
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Or maybe "This defines" rather than "This will create" since technically you don't create it until you run kubectl.

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+We may check the current status of autoscaler by running:
+
+```console
+$ kubectl get experimental/hpa
+NAME         REFERENCE                                   TARGET     CURRENT   MINPODS   MAXPODS   AGE
+php-apache   ReplicationController/default/php-apache/   100m cpu   0 cpu     1         10        4m
+```
+
+Please not that the current CPU consumption is 0 as we are not sending any requests to the server.
```
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: s/not/note/
Also, explain how the "current" column is computed (IIRC it's just the average across all the pods controlled by the corresponding RC?)

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+We may examine, how CPU load was increased (the results should be visible after about 2 minutes) by executing:
+
+```console
+$ kubectl get experimental/hpa
+NAME         REFERENCE                                   TARGET     CURRENT    MINPODS   MAXPODS   AGE
+php-apache   ReplicationController/default/php-apache/   100m cpu   471m cpu   1         10        10m
+```
+
+In the case presented here, it bumped CPU consumption to 471 milli-cores.
+In affect, the replication controller was resized to 5 replicas:
```
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: s/In affect/As a result/

BTW, in the next version of the code it would probably be good to add REPLICAS column to "get experimental/hpa" -- it is already in the HorizontalPodAutoscalerStatus so this should be trivial.

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+$ kubectl get experimental/hpa
+NAME         REFERENCE                                   TARGET     CURRENT   MINPODS   MAXPODS   AGE
+php-apache   ReplicationController/default/php-apache/   100m cpu   0 cpu     1         10        26m
+
+$ kubectl get rc
+CONTROLLER   CONTAINER(S)   IMAGE(S)                    SELECTOR         REPLICAS   AGE
+php-apache   php-apache     gcr.io/dev-jsz/php-apache   run=php-apache   1          42m
+```
+
+As we see, in the presented case CPU load drop to 0, and the number of replicas dropped to 1.
```
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: s/drop/dropped/

Also, you might want to use "utilization" instead of "load"

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 10 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 12 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 13 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 14 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 15 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 17 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 18 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 16 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

**Comment by jszczepkowski** [comment]
@davidopp Greate thanks for your comments! All are applied now. PTAL.

**Comment by fgrzadkowski** [comment]
@k8s-bot test this please

**Comment by fgrzadkowski** [comment]
@k8s-bot test this please

**Comment by HaiyangDING** [comment]
CC @HaiyangDING

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 9 above)
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: +1, please do this.

**Comment by davidopp** [comment]
Looks great except for the one comment from @fgrzadkowski which you didn't address yet (I added a +1 to it).

File: examples/experimental/horizontal-pod-autoscaling/README.md [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Horizontal Pod Autoscaler
+
+Horizontal pod autoscaling is an experimental feature in Kubernetes 1.1.
+It allows the number of pods in a replication controller or deployment to scale automatically based on observed CPU or memory usage.
+<b>Please note that the current API is tentative and will be subject to change before a stable version is released.</b>
```
Comment (davidopp) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: IIUC we're actually going to put this into the beta API, so I would word this a bit less strongly.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 9 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: Done.

File: examples/experimental/horizontal-pod-autoscaling/README.md (hunk 19 above)
Comment (jszczepkowski) [inline examples/experimental/horizontal-pod-autoscaling/README.md]: The API currently is in experimental and most probably there will be some changes before moving it to beta. I will change the warning as soon as we are in beta.

**Comment by jszczepkowski** [comment]
@davidopp PTAL

**Comment by davidopp** [comment]
LGTM
